tests/test_mutations.py

```diff
@@ -472,40 +704,226 @@ def test_keep_mutation_parent_zero_rate(self):
+                assert map_position[index] <= left
+                assert right <= map_position[index + 1]
+                assert right <= edge.right
+                # Generate the mutations.
+                rate = self.rate_map.value[index]
+                mu = rate * (site_right - site_left) * branch_length
+                for _ in range(rng.poisson(mu)):
+                    position = rng.flat(site_left, site_right)
+                    if discrete_sites:
+                        position = np.floor(position)
+                    assert edge.left <= position
+                    assert position < edge.right
+                    time = rng.flat(branch_start, branch_end)
+                    site_pos = np.searchsorted([s['site'].position for s in self.sites],
+                                               position)
+                    if (site_pos == len(self.sites)
+                       or self.sites[site_pos]['site'].position != position):
+                        # new site
+                        ancestral_state = self.alleles[self.pick_allele()]
+                        site = tskit.SiteTableRow(position=position,
+                                                  ancestral_state=ancestral_state,
+                                                  metadata=b'')
+                        site_item = {'site': site,
+                                     'mutation_list': []}
+                        self.sites.insert(site_pos, site_item)
+                    else:
+                        site_item = self.sites[site_pos]
+                    mut_pos = np.searchsorted([-1 * m['time']
+                                               for m in site_item['mutation_list']],
+                                              -1 * time)
+                    mut = tskit.MutationTableRow(site=-1, node=edge.child,
+                                                 derived_state="Y",
+                                                 parent=tskit.NULL, metadata=b'')
+                    mutation_item = {'time': time,
+                                     'mutation': mut,
+                                     'new': True}
+                    site_item['mutation_list'].insert(mut_pos, mutation_item)
+                    # Sample from this to duplicate the behaviour in the C implementation
+                    rng.uniform_int(1)
+                index += 1
+                left = right
+
+    def pick_allele(self):
+        u = np.random.uniform()
+        j = 0
+        while u > self.alphabet_distribution[j]:
+            u -= self.alphabet_distribution[j]
+            j += 1
+        return j
+
+    def transition_allele(self, j):
+        u = np.random.uniform()
+        k = 0
+        while u > self.mutation_matrix[j][k]:
+            u -= self.mutation_matrix[j][k]
+            k += 1
+        return k
+
+    def choose_alleles(self, tables, rng):
+        assert tables.sites.num_rows == len(self.sites)
```

Comment:
This is obscure to me too I'm afraid. So, at this point we have a complete model in the tables, ancestral states have been chosen and any new mutations need to have states chosen. I think we'll need to build the mutation tree explicitly and then traverse down it, won't we? Otherwise, keeping track of what the states are and any constraints we have will be too hard (i.e., what if we have a bunch of old mutations that already have states assigned in there?). Building the mutation tree will be very cheap, doing those traversals will be fast and more importantly it'll be really easy to understand what's happening.

Reply:
Yep - having to check which sites/mutations need derived states generated and which were already in the tree sequence, and which will end up being no-op sites or mutations and thus removed has made this too complicated. I'm going to rewrite with a tree traversal: wait and see, it'll be much simpler.